### PR TITLE
[CUDA graphs] [JIT] Capture-safe RNG in nvfuser

### PR DESCRIPTION
**Pull request by mcarilli**
(formerly https://github.com/pytorch/pytorch/pull/50148, @csarofeen asked me to PR nvfuser diffs here first)

Eager mode RNG kernels needed some minor changes to interact safely with cuda graphs.  This PR extends those changes to the kernels generated by nvfuser.

One thing I'm unclear on is the best way to let NVRTC know the definition of PhiloxCudaState (defined in [ATen/CUDAGeneratorImpl.h](https://github.com/pytorch/pytorch/blob/74c055b24065d0202aecdf4bc837d3698d1639e1/aten/src/ATen/CUDAGeneratorImpl.h#L89)).  I suggested two options in comments ([1](https://github.com/csarofeen/pytorch/pull/593/files#diff-6f1e7ce30c1600288756ba97ab7b4324993d2959452d874f0d15e79a16e1f850R38-R40), [2](https://github.com/csarofeen/pytorch/pull/593/files#diff-5e44e2a0e9a686349e4152cd3f1255ae55c3c208b22112f194144d8c6612f61aR45-R49)) but im not sure.

Another thing I'm unclear on is the best way to test these diffs.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp]: this is the best option with what we have today. the new file under nvfuser_resources should be self-contained  (unfortunately it can't use `#include`).

Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp]: Uh oh, that means I either have to duplicate my struct definition from CUDAGeneratorImpl.h manually (brittle) or have something autogen the duplication (complicated).
out of curiosity why can't it use #include?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp]: yes, you'd need to duplicate the definition. besides the options you mentioned, another idea would be to have the canonical definition under nvfuser_resources, and #include that one from somewhere else - I mention this for brainstorming only, as I think it would be a hack and I don't recommend it.

`#include` doesn't work since we are missing the mechanisms to setup the nvrtc include locations (we prototyped this and it can be done, but upstream maintainers had reservations - I still hope we'll overcome the objections at some point but we're not there yet). 

the alternative currently implemented is a custom preprocessor for the files under nvfuser_resources, which generates string literals from the .cu files. Technically it would be possible to do a form of C preprocessing at that point, although that would introduce some obvious (and maybe less obvious) complications.

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp]: actually, we may have a better solution: if the common code is already in a standalone file we could give it the same treatment as we do for files under nvfuser_resources. The only constraint is that the common file should be valid for textual insertion in the kernel "preamble" - ex. it can't have #includes itself

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: does this have to be in that at namespace? this could lead to subtle odr violations which would likely be hard to troubleshoot

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: not needed

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: I'd initialize all the members here

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: same question as the at namespace

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: we should get rid of std::tuple here (create a dedicated structure instead)

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: if we go down the copy-and-paste route, this warning should accompany both definitions

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: I guess not, but I don't understand why it's a problem.  Classes are always defined in headers without odr violation as long as member functions in the header are inline.  Why is it a problem in this case?

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: I think I wrote warnings at all definition locations, do you see one i missed?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: I only looked at one commit, my bad. I see that the warning is in CUDAGeneratorImpl.h too.

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu:114]: Is it because this stuff will become a distinct .so from the similarly-defined stuff in ATen?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: the headers generally work since the definitions are identical. If the definitions diverge, even in seemingly harmless ways then it's undefined what happens - in practice the issues are hard to troubleshoot since there's no toolchain support today.

PS. even with headers you can get ODR issues if the translation units are compiled with different flags or with different macros

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: > Is it because this stuff will become a distinct .so from the similarly-defined stuff in ATen?

No, that's an ABI compatibility thing. The potential problem is if anyone ends up including both the aten header and the nvfuser_resources - it seems a bit stretched in the current setup, but let's say someone wants to prototype some CUDA changes by hand. Or if we end up adding support for real header includes to hvrtc.

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: Eager mode kernels always used std::tuple<uint64_t, uint64_t>, even before my changes.  What's the problem?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: 1. unnecessary dependency on the standard library
2. using a specialized struct results in better ergonomics, type safety and readability

#2 is the big one, although for nvrtc #1 is also something to keep in mind

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: If CPU-compiled and nvrtc-compiled instances of PhiloxCudaState are ABI-incompatible, aren't we screwed regardless because CPU instances are bitcopied into the kernel instances by cuLaunchKernel?  How does putting the definition nvrtc compiles in a different namespace save us?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: it's the other way around: putting things in the same namespace may lead to extra issues even if we're layout/abi compatible.

File: aten/src/ATen/CUDAGeneratorImpl.h
Comment (tlemo) [inline aten/src/ATen/CUDAGeneratorImpl.h]: if we use this refactoring strategy (which I think it's much better than copy & paste) - then I we should make the shared headers be as self contained as possible. So move the `namespace at` into PhiloxCudaStateRaw.cuh instead of depending on the surrounding context.

Comment (mcarilli) [reply aten/src/ATen/CUDAGeneratorImpl.h]: Done, but how will that play with your earlier concerns about namespaces?  Tbh im not sure how exactly you wanted me to organize things.

Comment (tlemo) [reply aten/src/ATen/CUDAGeneratorImpl.h]: With copy-and-paste, you want the minimal contract - even small source code changes can have unexpected consequences so you want compatible but different definitions.

With a shared "header" we can afford the same namespace. (we can still use different namespaces, but it's not a requirement to do so). Since we can afford same namespace, then the problem becomes packaging the shared file in a robust and intuitive way. We should factor the shared definition as a regular header - which means it shouldn't not depend much on the context where it's included. 

Does this make sense?

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: done, but this change bleeds significantly into eager mode kernels.  It probably won't cause merge conflicts the next time you pull in upstream, but it might.

File: aten/src/ATen/cuda/detail/PhiloxCudaStateRaw.cuh
Comment (tlemo) [inline aten/src/ATen/cuda/detail/PhiloxCudaStateRaw.cuh]: I'd add a comment here explaining why this "header" doesn't have an include guard (or #pragma once)

File: aten/src/ATen/cuda/detail/UnpackRaw.cuh
Comment (tlemo) [inline aten/src/ATen/cuda/detail/UnpackRaw.cuh]: ditto

File: torch/csrc/jit/codegen/cuda/codegen.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/codegen.cpp]: why do we need this #include ?

File: torch/csrc/jit/codegen/cuda/executor_kernel_arg.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor_kernel_arg.h]: go for it!

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu:1]: why was this line removed?

**Review by tlemo — APPROVED** [review]
LGTM with a few small comments

File: torch/csrc/jit/codegen/cuda/codegen.cpp
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/codegen.cpp]: You're right, we don't need it.  I instinctively included it because I mention PhiloxCudaState in the file, but all we're doing in the file is building strings.

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu
Comment (mcarilli) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu:1]: no reason, it was random whitespace at the top of the file.

**Review by csarofeen — APPROVED** [review]
LGTM

**Comment by mcarilli** [comment]
Don't merge this yet.

After discussions with Christian, I'm going to split all the non-nvfuser diffs into a separate PR against upstream.  They are simple so it should land quickly. The nvfuser diffs will stay in this PR, and because they require the non-nvfuser diffs, we'll keep this PR in limbo until the non-nvfuser diffs are pulled back from upstream into 20_12_3_devel.

Ensuring all non-nvfuser diffs land upstream first alleviates the need for Facebook to review any non-nvfuser diffs when the big merge of our devel branch into upstream eventually happens.

**Comment by mcarilli** [comment]
I moved all the diffs outside torch/csrc/jit to https://github.com/pytorch/pytorch/pull/51580.

This PR can be merged after https://github.com/pytorch/pytorch/pull/51580 lands upstream and we pull its changes back in.

**Comment by jjsjann123** [comment]
Tests passed merging this one.

**Pull request merged**
